### PR TITLE
feat: add model tag restatement selector

### PR DESCRIPTION
**Pull request by eakmanrq**

### CODE REVIEW

File: sqlmesh/core/context.py [hunk 1]
```diff
@@ -867,6 +874,14 @@ def plan(
         if restate_models is not None:
             restate_models = model_selector.expand_model_selections(restate_models)
 
+        if restate_tags is not None:
+            restate_models = restate_models or set()
+            restate_models.update(model_selector.expand_model_tags(restate_tags))
+            if not restate_models:
+                raise PlanError(
+                    f"No models match the provided tags. Tags: {', '.join(restate_tags)}"
+                )
```
Comment (eakmanrq) [inline sqlmesh/core/context.py:883]: Do we think we should raise here?

File: sqlmesh/core/selector.py [hunk 2]
```diff
@@ -108,35 +178,29 @@ def expand_model_selections(self, model_selections: t.Iterable[str]) -> t.Set[st
...
+                selection,
+                include_upstream,
+                include_downstream,
+            ) = self._get_value_and_dependency_inclusion(selection)
 
             if "*" in selection:
                 for model in self._models.values():
-                    if fnmatch.fnmatch(model.name, selection):
-                        _add_model(model.fqn, include_upstream, include_downstream)
+                    if fnmatch.fnmatchcase(model.name, selection):
```
Comment (eakmanrq) [inline sqlmesh/core/selector.py:193]: @izeigerman I think you wanted to use `fnmatchcase` here instead of `fnmatch`. `fnmatch` determines if the matching is case-sensitive based on the OS which feels irrelevant within this context. Therefore I think always matching on case makes more sense here.

File: mkdocs.yml [hunk 3]
```diff
@@ -105,6 +105,7 @@ theme:
         icon: material/weather-night
         name: Switch to light mode
   features:
+    - content.tabs.link
```
Comment (eakmanrq) [inline mkdocs.yml:108]: This makes it so that if a user picks an option on a page (like "Model" or "Tag" in this example) then it updates all of the codeblocks to match that selection.

**Comment by izeigerman** [comment]
Instead of adding new arguments like `--restate-tag`, `--select-tag`, etc., can we instead extend the model selection syntax to incorporate tags? Eg. `--select-model tag:my_tag`?

File: sqlmesh/core/context.py (hunk 1 above)
Comment (izeigerman) [reply sqlmesh/core/context.py:883]: Should this just be a noop instead?

**Comment by eakmanrq** [comment]
> Rather than adding new arguments like `--restate-tag`, `--select-tag`, etc., can we instead extend the model selection syntax to incorporate tags? Eg. `--select-model tag:my_tag`?

Great suggestion will do.

File: sqlmesh/core/context.py [hunk 4]
```diff
@@ -877,12 +882,14 @@ def plan(
                 environment or c.PROD,
                 snapshots=self._snapshots(models_override),
                 create_from=create_from,
+                force_no_diff=(restate_models is not None and not expanded_restate_models)
```
Comment (izeigerman) [inline sqlmesh/core/context.py:885]: Why do we need this?

File: sqlmesh/core/selector.py [hunk 5]
```diff
@@ -97,7 +113,67 @@ def select_models(
 
         return models
 
-    def expand_model_selections(self, model_selections: t.Iterable[str]) -> t.Set[str]:
+    @classmethod
```
Comment (izeigerman) [inline sqlmesh/core/selector.py:116]: maybe `@staticmethod` instead

File: sqlmesh/core/selector.py [hunk 6]
```diff
@@ -106,37 +182,38 @@ def expand_model_selections(self, model_selections: t.Iterable[str]) -> t.Set[st
...
-                    include_upstream,
-                    include_downstream,
-                )
+                model_fqn = normalize_model_name(selection, self._default_catalog, self._dialect)
+                if model_fqn in models:
+                    matched_models.add(model_fqn)
 
-        return result
+            if not matched_models:
+                logger.warning(f"Model {selection} doesn't match any models.")
```
Comment (izeigerman) [inline sqlmesh/core/selector.py:214]:
```suggestion
                logger.warning(f"Expression '{selection}' doesn't match any models.")
```

File: sqlmesh/core/selector.py [hunk 7]
```diff
@@ -97,7 +113,67 @@ def select_models(
...
+
+        if "*" in selection:
+            for model_tag in self._models_by_tag:
+                if fnmatch.fnmatchcase(model_tag, selection):
+                    matched_tags.add(model_tag)
+        elif selection in self._models_by_tag:
+            matched_tags.add(selection)
+
+        if not matched_tags:
+            logger.warning(f"Tag {tag_selection} doesn't match any models.")
```
Comment (izeigerman) [inline sqlmesh/core/selector.py:166]:
```suggestion
            logger.warning(f"Expression '{tag_selection}' doesn't match any models.")
```

**Comment by izeigerman** [comment]
@eakmanrq  Just to confirm: this update doesn't cover using a superset of models from both the local state and the target environment?

**Comment by eakmanrq** [comment]
@izeigerman Is this what you are referring to? https://github.com/TobikoData/sqlmesh/blob/e30e60a5a7e157648e6ae2b81f5ac5f413429b7f/sqlmesh/core/selector.py#L89-L91

**Comment by izeigerman** [comment]
@eakmanrq yes, thank you, overlooked that

**Review by izeigerman — APPROVED** [review]

File: sqlmesh/core/context.py (hunk 4 above)
Comment (eakmanrq) [reply sqlmesh/core/context.py:885]: If you don't do this then you will have a diff since you are including snapshots and other configuration in plan.

File: sqlmesh/core/selector.py [hunk 8]
```diff
@@ -97,7 +113,67 @@ def select_models(
...
+
+        if "*" in selection:
+            for model_tag in self._models_by_tag:
+                if fnmatch.fnmatchcase(model_tag, selection):
+                    matched_tags.add(model_tag)
+        elif selection in self._models_by_tag:
+            matched_tags.add(selection)
+
+        if not matched_tags:
+            logger.warning(f"Expression '{tag_selection}' doesn't match any models.")
```
Comment (izeigerman) [inline sqlmesh/core/selector.py:166]:
```suggestion
            logger.warning(f"Expression 'tag:{tag_selection}' doesn't match any models.")
```
Sorry, I've just realized that you strip the `tag:` prefix upstream.
